File: 03-poo/13-excepciones-personalizadas/scala/soluciones/RetoSaldo.scala
class SaldoInsuficienteReto(saldo: Double, monto: Double) extends Exception(s"Saldo $saldo menor que retiro $monto")
class CuentaReto(private var saldo: Double):
  def retirar(monto: Double): Unit =
    if monto > saldo then throw new SaldoInsuficienteReto(saldo,monto)
    saldo -= monto

@main def retoSaldo(): Unit =
  try new CuentaReto(100).retirar(120)
  catch case e: SaldoInsuficienteReto => println(e.getMessage)

File: 03-poo/13-excepciones-personalizadas/java/soluciones/RetoSaldo.java
public class RetoSaldo {
    static class SaldoInsuficienteException extends Exception { SaldoInsuficienteException(double saldo,double monto){ super("Saldo "+saldo+" menor que retiro "+monto); } }
    static class Cuenta { private double saldo; Cuenta(double saldo){this.saldo=saldo;} void retirar(double monto) throws SaldoInsuficienteException { if(monto>saldo) throw new SaldoInsuficienteException(saldo,monto); saldo-=monto; } }
    public static void main(String[] args){ try{ new Cuenta(100).retirar(120); }catch(SaldoInsuficienteException e){ System.out.println(e.getMessage()); } }
}
